File: java/src/main/java/com/marcusmccurdy/trie/Trie.java
package com.marcusmccurdy.trie;

import java.util.*;
import java.util.Map.Entry;

/**
 *
 * @author Marcus McCurdy <marcus.mccurdy@gmail.com>
 */
public class Trie {

    protected final Map<Character, Trie> children;
    protected String terminal;
    protected boolean leaf = false;

    public Trie() {
        this(null);
    }

    private Trie(String value) {
        this.terminal = value;
        children = new HashMap<Character, Trie>();
    }

    protected void add(char c) {
        String val;
        if (this.terminal == null) {
            val = Character.toString(c);
        } else {
            val = this.terminal + c;
        }
        children.put(c, new Trie(val));
    }

    public void insert(String word) {
        if (word == null) {
            throw new IllegalArgumentException("Cannot add null to a Trie");
        }
        Trie node = this;
        for (char c : word.toCharArray()) {
            if (!node.children.containsKey(c)) {
                node.add(c);
            }
            node = node.children.get(c);
        }
        node.leaf = true;
    }

    public String find(String word) {
        Trie node = this;
        for (char c : word.toCharArray()) {
            if (!node.children.containsKey(c)) {
                return "";
            }
            node = node.children.get(c);
        }
        return node.terminal;
    }

    public Collection<String> autoComplete(String prefix) {
        Trie node = this;
        for (char c : prefix.toCharArray()) {
            if (!node.children.containsKey(c)) {
                return Collections.emptyList();
            }
            node = node.children.get(c);
        }
        return node.allPrefixes();
    }

    protected Collection<String> allPrefixes() {
        List<String> results = new ArrayList<String>();
        if (this.leaf) {
            results.add(this.terminal);
        }
        for (Entry<Character, Trie> entry : children.entrySet()) {
            Trie child = entry.getValue();
            Collection<String> childPrefixes = child.allPrefixes();
            results.addAll(childPrefixes);
        }
        return results;
    }
}

File: trie.go
package trie

import (
	"errors"
)

type Trie struct {
	Value    string
	Children []Child
	End      bool
}

type Child struct {
	Value rune
	Link  *Trie
}

func NewTrie() *Trie {
	return &Trie{}
}

func NewChild(value rune) Child {
	child := Child{}
	child.Value = value
	return child
}

func (t *Trie) Add(value rune) *Trie {
	link := NewTrie()
	link.Value = t.Value + string(value)
	child := NewChild(value)
	child.Link = link

	t.Children = append(t.Children, child)

	return link
}

func (t *Trie) Insert(value string) {
	trie := t

	for i, val := range value {
		link, err := trie.Find(val)

		if i < len(value) {
			if err != nil {
				link = trie.Add(val)
			}

			trie = link
		}

	}

	trie.End = true
}

func (t *Trie) Find(value rune) (*Trie, error) {
	if t != nil {
		for _, child := range t.Children {
			if child.Value == value {
				return child.Link, nil
			}
		}
	}
	return nil, errors.New("Value not found")
}

func (t *Trie) AllPrefixes() ([]string, error) {
	results := []string{}
	if t != nil {
		if t.End == true {
			results = append(results, t.Value)
		}

		for _, val := range t.Children {
			prefixes, err := val.Link.AllPrefixes()
			if err != nil {
				break
			}
			results = append(results, prefixes...)
		}
	} else {
		return results, errors.New("No end found, so no completions available")
	}
	return results, nil
}

func (t *Trie) AutoComplete(prefix string) ([]string, error) {
	trie := t

	for _, val := range prefix {
		link, _ := trie.Find(val)
		trie = link
	}

	results, err := trie.AllPrefixes()

	return results, err
}
